File: GameEngine_PC/GameEngineOpenGL3_2/src/main/java/gameEngine/models/GuiShape.java
package gameEngine.models;

import com.dferreira.commons.shapes.IExternalMaterial;
import com.dferreira.commons.shapes.IShape;

/**
 * Represents the 2D GUI
 */
public class GuiShape implements IShape {
	private static final float SIZE = 1f;

	/**
	 * Vertices of the GUI quad
	 */
	private static final float[] vertices = { -SIZE, SIZE, -SIZE, -SIZE, SIZE, SIZE, SIZE, -SIZE };

	/**
	 * Constructor of the GUI shape
	 */
	public GuiShape() {
	}

	/**
	 * 
	 * @return The vertices of the quad
	 */
	@Override
	public float[] getVertices() {
		return vertices;
	}

	/**
	 * @return the Coordinates of the textures of the shape
	 */
	@Override
	public float[] getTextureCoords() {
		return null;
	}

	/**
	 * @return the normal vectors that make the shape
	 */
	@Override
	public float[] getNormals() {
		return null;
	}

	/**
	 * @return The indices of the vertices that make the shape
	 */
	@Override
	public int[] getIndices() {
		return null;
	}

    /**
     * @return The group that the gui shape belongs (if any)
     */
	@Override
	public String getGroupName() {
		return null;
	}

	@Override
	public IExternalMaterial getMaterial() {
		return null;
	}

}

File: GameEngine_IOS/GameEngine/Models/GuiShape.swift
import Foundation

/**
 * Represents the 2D GUI
 */
public class GuiShape : NSObject, IShape {
    private var _vertices : UnsafeMutablePointer<Float> ;
    
    public static let SIZE : Float = 1.0
    private let GUI_NUMBER_OF_ELEMENTS : Int = 8;
    
    /**
     * Vertices of the GUI quad
     */
    private let guiVertexData : Array<Float> = [
        -SIZE, SIZE, -SIZE, -SIZE, SIZE, SIZE, SIZE, -SIZE
    ];
    
    
    /**
     * Initiator of the GUI shape
     */
    public override init() {
        //Allocate and fill the vertices memory
        self._vertices = UnsafeMutablePointer<Float>.alloc(guiVertexData.count);
        
        //Copy vertices one by one
        for i in 0 ..< guiVertexData.count {
            self._vertices[i] = guiVertexData[i];
        }
    }
    
    /**
     * Deinitialization of the gui shape
     */
    deinit {
        free(self._vertices);
    }
    
    
    /**
     *
     * @return The vertices of the quad
     */
    public func getVertices() -> UnsafeMutablePointer<Float> {
        return self._vertices;
    }
    
    /**
     * @return number of vertices that make the shape
     */
    public func countVertices() -> Int {
        return GUI_NUMBER_OF_ELEMENTS;
    }
    
    /**
     * @return the Coordinates of the textures of the shape
     */
    public func getTextureCoords()  -> UnsafeMutablePointer<Float> {
        return nil;
    }
    
    /**
     *Number of the texture coordinates
     */
    public func countTextureCoords() -> Int {
        return 0;
    }
    
    /**
     * @return the normal vectors that make the shape
     */
    public func getNormals() -> UnsafeMutablePointer<Float> {
        return nil;
    }
    
    /*
     * Number of normal that the shape has
     */
    public func countNormals() -> Int {
        return 0;
    }
    
    /**
     * @return The indices of the vertices that make the shape
     */
    public func getIndices() -> UnsafeMutablePointer<ushort> {
        return nil;
    }
    
    /*
     Number of indices that the shape has
     */
    public func countIndices() -> Int {
        return 0;
    }
    
    /**
     * @return the groupName Name of the group wish belongs
     */
    public func getGroupName() -> String? {
        return nil;
    }
    
    /**
     *
     * @return The material that is associated with shape
     */
    public func getMaterial() -> IExternalMaterial? {
        return nil;
    }
}
